### PR TITLE
Add support for subject alternative name

### LINKED ISSUE #4: Allow Subject Alternative Names in certificates
multiple domain in the same certificate

See https://github.com/acmephp/core/pull/4

Don't forget to:
- play challenge for each alternative name
- in case a previous certificate exists, open and read it to chek if new domains need a challenge

### PR DESCRIPTION
**Pull request by jderusse**
Fix #4

### CODE REVIEW

File: Acme/Certificate/Extractor/CertificateExtractor.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+use AcmePhp\Bundle\Exception\ParsingCertificateException;
+use AcmePhp\Core\Ssl\Certificate;
+
+/**
+ * Extract data from certificate file.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class CertificateExtractor implements ExtractorInterface
```
Comment (tgalopin) [inline Acme/Certificate/Extractor/CertificateExtractor.php]: What about "reader"? I'd fine it more intuitive.

File: Acme/Certificate/Extractor/CertificateExtractor.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    public function extract($content)
+    {
+        $rawData = openssl_x509_parse($content);
+        if (!$rawData) {
+            throw new ParsingCertificateException(
+                sprintf('Certificate parsing failed with error: %s', openssl_error_string())
+            );
+        }
+
+        return [
```
Comment (tgalopin) [inline Acme/Certificate/Extractor/CertificateExtractor.php]: What do you think of creating a read only value object for that structure? It seems logical and let ide autocomplete

File: Acme/Domain/Challenger.php [hunk 3]
```diff
@@ -40,19 +40,32 @@ public function __construct(AcmeClient $client, EventDispatcherInterface $dispat
...
-     * Request and check a domain.
+     * Request and check a list of domain.
      *
      * @param DomainConfiguration $configuration
      *
      * @throws \Exception
      */
-    public function challengeDomain(DomainConfiguration $configuration)
+    public function challengeDomains(array $domains)
     {
```
Comment (tgalopin) [inline Acme/Domain/Challenger.php:51]: Can you check domains using Assert?

File: Acme/Certificate/Extractor/CertificateExtractor.php (hunk 2 above)
Comment (jderusse) [inline Acme/Certificate/Extractor/CertificateExtractor.php]: You're right, I was too lazy

File: Acme/Certificate/Extractor/CertificateExtractor.php (hunk 1 above)
Comment (jderusse) [inline Acme/Certificate/Extractor/CertificateExtractor.php]: Imo, it read nothing. Maybe parser is a proper word.

But I was thinking to extractors like in https://github.com/symfony/symfony/tree/master/src/Symfony/Component/PropertyInfo/Extractor

File: Acme/Domain/Challenger.php (hunk 3 above)
Comment (jderusse) [inline Acme/Domain/Challenger.php:51]: Yup

File: Acme/Certificate/Extractor/CertificateExtractor.php (hunk 1 above)
Comment (tgalopin) [inline Acme/Certificate/Extractor/CertificateExtractor.php]: I would prefer Parser indeed. For me, a certificate extractor extract the certificate from somewhere, and that's not the aim of this class.

**Comment by jderusse** [comment]
Fixed

File: Acme/Certificate/CertificateMetadata.php [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+namespace AcmePhp\Bundle\Acme\Certificate;
+
+use AcmePhp\Core\Ssl\Certificate;
+
+/**
+ * Entity representing a certificate.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class CertificateMetadata
```
Comment (tgalopin) [inline Acme/Certificate/CertificateMetadata.php:21]: I think this entity can be read-only. It makes no sense to change it once a certificate was parsed, does it?

**Comment by tgalopin** [comment]
Other than that, it's great, thanks :) !

File: Acme/Certificate/CertificateMetadata.php (hunk 4 above)
Comment (jderusse) [inline Acme/Certificate/CertificateMetadata.php:21]: Do you think about the merge function ? or Setters ?

For the first one, I may replace it by a static function which return a new immutable instance (does it worst it ?)
For the second, it's a mistake, I'll fix it

Comment (jderusse) [inline Acme/Certificate/CertificateMetadata.php:21]: You're right, it's overcomplecated to have several parsers. I'll refactor it

Comment (tgalopin) [inline Acme/Certificate/CertificateMetadata.php:21]: I was talking about setters, but I thought the same: a single parser is enough IMO, so the merge function could be removed.

Comment (jderusse) [inline Acme/Certificate/CertificateMetadata.php:21]: done

Comment (jderusse) [inline Acme/Certificate/CertificateMetadata.php:21]: fixed. I remove the parseInterface thing

**Pull request merged**